synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -125,28 +110,42 @@ def __init__(self, hs: "HomeServer"):
         # Whether to support MSC3772 is supported.
         self._relations_match_enabled = self.hs.config.experimental.msc3772_enabled
 
+    @lru_cache()
+    async def get_rules(self, event: EventBase) -> Dict[str, List[Dict[str, dict]]]:
+        """Given an event return the rules for all users who are
+        currently in the room.
+        """
+
+        local_users = await self.store.get_local_users_in_room(event.room_id)
```

Comment:
We'll want to invalidate the functions cache if `get_local_users_in_room` or `bulk_get_push_rules` cache is invalidated. We can do that using the `cache_context` machinery by a) adding a `cache_context` parameter to the function, and b) calling this with `on_invalidate=cache_context.invalidate`.

An example of this in action:

https://github.com/matrix-org/synapse/blob/d102ad67fddc650c34baa89dc7b2926d46a9aeca/synapse/storage/databases/main/appservice.py#L153-L163